mypyc/lib-rt/CPyFunction.c

```diff
@@ -0,0 +1,235 @@
+}
+
+static PyType_Slot CPyFunction_slots[] = {
+    {Py_tp_dealloc, (void *)CPyFunction_dealloc},
+    {Py_tp_repr, (void *)CPyFunction_repr},
+    {Py_tp_call, (void *)CPyFunction_call},
+    {Py_tp_traverse, (void *)CPyFunction_traverse},
+    {Py_tp_clear, (void *)CPyFunction_clear},
+    {Py_tp_members, (void *)CPyFunction_members},
+    {Py_tp_getset, (void *)CPyFunction_getsets},
+    {Py_tp_descr_get, (void *)CPy_PyMethod_New},
+    {0, 0},
+};
+
+static PyType_Spec CPyFunction_spec = {
+    .name = "Function compiled with mypyc",
+    .basicsize = sizeof(CPyFunction),
+    .itemsize = 0,
+    .flags =
+#if PY_VERSION_HEX >= 0x030A0000
+             Py_TPFLAGS_IMMUTABLETYPE |
+#endif
+#if PY_VERSION_HEX >= 0x030C0000
+             Py_TPFLAGS_MANAGED_DICT |
+#endif
+             Py_TPFLAGS_HAVE_VECTORCALL | Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC | Py_TPFLAGS_BASETYPE,
+    .slots = CPyFunction_slots,
+};
+
+static PyTypeObject *CPyFunctionType = NULL;
+
+static PyObject* CPyFunction_Vectorcall(PyObject *func, PyObject *const *args, size_t nargsf, PyObject *kwnames) {
+    CPyFunction *f = (CPyFunction *)func;
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    PyObject *self;
+    PyCFunction meth = ((PyCFunctionObject *)f)->m_ml->ml_meth;
+
+    self = ((PyCFunctionObject *)f)->m_self;
+    return ((_PyCFunctionFastWithKeywords)(void(*)(void))meth)(self, args, nargs, kwnames);
+}
+
+
+static CPyFunction* CPyFunction_Init(CPyFunction *op, PyMethodDef *ml, PyObject* name,
+                                     PyObject *module, PyObject* code) {
+    PyCFunctionObject *cf = (PyCFunctionObject *)op;
+    CPyFunction_weakreflist(op) = NULL;
+    cf->m_ml = ml;
+    cf->m_self = (PyObject *) op;
+
+    Py_XINCREF(module);
+    cf->m_module = module;
+
+    Py_INCREF(name);
+    op->func_name = name;
+
+    ((PyCMethodObject *)op)->mm_class = NULL;
+
+    Py_XINCREF(code);
+    op->func_code = code;
+
+    CPyFunction_func_vectorcall(op) = CPyFunction_Vectorcall;
+    return op;
+}
+
+static PyObject* CPyCode_New(const char *filename, const char *funcname, int first_line, int flags) {
+    PyCodeObject *code_obj = PyCode_NewEmpty(filename, funcname, first_line);
+    code_obj->co_flags = flags;
+    return (PyObject *)code_obj;
+}
+
+static PyMethodDef* CPyMethodDef_New(const char *name, PyCFunction func, int flags, const char *doc) {
+    PyMethodDef *method = (PyMethodDef *)PyMem_Malloc(sizeof(PyMethodDef));
+    method->ml_name = name;
+    method->ml_meth = func;
+    method->ml_flags = flags;
+    method->ml_doc = doc;
+    return method;
+}
+
+PyObject* CPyFunction_New(PyObject *module, const char *filename, const char *funcname,
+                          PyCFunction func, int func_flags, const char *func_doc,
+                          int first_line, int code_flags) {
+    if (!CPyFunctionType) {
+        CPyFunctionType = (PyTypeObject *)PyType_FromSpec(&CPyFunction_spec);
+        assert(CPyFunctionType);
```

Comment:
If this fails, return `NULL` instead of failing an assertion. Another option, since this will likely only fail if running out of memory, is to call `CPyError_OutOfMemory` on NULL return which aborts.